### PR TITLE
Aida 293

### PR DESCRIPTION
**Pull request by gitnordt**

### CODE REVIEW

File: src/app/components/base-neon-component/base-neon.component.ts [hunk 1]
```diff
@@ -1140,4 +1140,31 @@ export abstract class BaseNeonComponent implements OnInit, OnDestroy {
...
     }
+
+    /**
+     * Dynamic sorting over an array of objects
+     * https://www.sitepoint.com/sort-an-array-of-objects-in-javascript/
+     *
+     * @arg {number} item
+     * @return {string}
+     */
+    compareValues(key: string, order: string = 'asc') {
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/base-neon-component/base-neon.component.ts]: Should this function be moved into the neonUtilities?  https://github.com/NextCenturyCorporation/neon-dash-internal/blob/master/src/app/neon-namespaces.ts#L33

File: src/app/components/network-graph/network-graph.component.ts [hunk 2]
```diff
@@ -871,23 +884,39 @@ export class NetworkGraphComponent extends BaseNeonComponent implements OnInit,
...
 
-                        for (const nodeLabel of this.prettifiedNodeLabels) {
-                            if (nodeLabel === shortName) {
-                                let colorMapVal = nodeColorField && nodeLabel;
-                                linkColor = this.colorSchemeService.getColorFor(nodeColorField, colorMapVal).toRgb();
-                                break;
+            if (links) {
+                for (let j = 0; j < links.length && graph.nodes.length < limit; j++) {
+                    let linkEntry = links[j];
+                   let linkNode  = this.allData.find((item) => item.kbid === linkEntry);
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/network-graph/network-graph.component.ts]: Looks like the indentation is off here?

File: src/app/components/base-neon-component/base-neon.component.ts (hunk 1 above)
Comment (ThomasSchellenbergNextCentury) [reply src/app/components/base-neon-component/base-neon.component.ts]: Also, are there tests for this function?

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.html [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+                <mat-form-field>
+                    <mat-select placeholder="Child Field" [(ngModel)]="options.typeField" required="false" (ngModelChange)="handleChangeData()" [disabled]="options.fields.length < 1">
+                        <mat-option [value]="emptyField">(None)</mat-option>
+                        <mat-option *ngFor="let field of options.fields" [value]="field">{{ field.prettyName }}</mat-option>
+                    </mat-select>
+                </mat-form-field>
+                <p></p>
+
+                <!--             TODO: Implement sub-child
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.html]: Is there a ticket for this TODO?  If so, please add the ticket number in the TODO comment.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.html [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                </mat-form-field>
+                <p></p>
+
+                <mat-form-field>
+                    <mat-select placeholder="Child Field" [(ngModel)]="options.typeField" required="false" (ngModelChange)="handleChangeData()" [disabled]="options.fields.length < 1">
+                        <mat-option [value]="emptyField">(None)</mat-option>
+                        <mat-option *ngFor="let field of options.fields" [value]="field">{{ field.prettyName }}</mat-option>
+                    </mat-select>
+                </mat-form-field>
+                <p></p>
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.html:70]: Please add selects or toggles for the other taxonomy options:  ascending, ignoreSelf, idField, linkField, filterFields.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+    /**
+     * Returns whether the data and fields for the visualization are valid.
+     *
+     * @return {boolean}
+     * @override
+     */
+    isValidQuery(): boolean {
+        return !!(this.options.database && this.options.database.name && this.options.table && this.options.table.name &&
+            this.options.idField && this.options.idField.columnName && this.options.categoryField &&
+            this.options.categoryField.columnName && this.options.typeField && this.options.typeField.columnName);
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Are categoryField and typeField actually required?  If so, please mark them as required in their mat-selects and remove the emptyField option in the HTML.  If they're not required, please remove them from isValidQuery here.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+        if (!this.filterExists(filter.field, filter.value)) {
+            this.filters.push(filter);
+            this.addNeonFilter(runQuery, filter, clause);
+        }
+
+    }
+
+    handleChangeData() {
+        super.handleChangeData();
+    }
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: You can just remove this function (you don't need to overwrite the superclass version).

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+     * Handles any post-initialization behavior needed with properties for the visualization.
+     *
+     * @override
+     */
+    postInit() {
+        this.executeQueryChain();
+    }
+
+    onEvent = ($event) => {
+        //needed for the taxonomy options
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Why is this needed?  Does this event trigger something else too?

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+    }
+
+    /**
+     * Removes the given visualization filter object from this visualization.
+     *
+     * @arg {object} filter
+     * @override
+     */
+    removeFilter(filter: any) {
+        //TODO:Update Taxonomy when filter from filter tray is deleted
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Please make a ticket for this bug and add the ticket number in the comment here.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+            }
+        }
+    }
+
+    /**
+     * Called after the filters in the filter service have changed.
+     * Defaults to calling setupFilters() then executeQueryChain()
+     */
+    handleFiltersChangedEvent(): void {
+        //TODO:Update Taxonomy when filter from filter tray is deleted
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Also add the ticket number for fixing this bug to the comment here too.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+
+            for (let child of node.children) {
+                if (!child.data.checked) {
+                    allChildChecked = false;
+                } else if (child.data.checked) {
+                    noChildChecked = false;
+                }
+            }
+
+//todo: toggling children causes indeterminate to turn into checked
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Please make a ticket for this bug and add the ticket number in the comment here.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+import { FieldMetaData } from '../../dataset';
+import { neonUtilities, neonVariables } from '../../neon-namespaces';
+import * as neon from 'neon-framework';
+import { current } from 'codelyzer/util/syntaxKind';
+
+/**
+ * Manages configurable options for the specific visualization.
+ */
+export class TaxonomyViewerOptions extends BaseNeonOptions {
+    // TODO Add and remove properties as needed.  Do NOT assign defaults to fields or else they will override updateFieldsOnTableChanged.
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: You can remove this comment.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+
+                        //If the parent(category) node does not exist in the tree, add it
+                        if (!foundCategory) {
+                        let parent = {
+                            id: counter++,
+                            name: category,
+                            children: [],
+                            description: this.options.categoryField.columnName,
+                            checked: true
+                        };
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: It looks like some of the indentation in this function got messed up somehow...

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 13]
```diff
@@ -0,0 +1,690 @@
...
+                                                    let subTypeObject = {
+                                                        id: counter++,
+                                                        name: type,
+                                                        lineage: category,
+                                                        description: this.options.subTypeField.columnName,
+                                                        checked: true
+                                                    };
+                                                    //Alphabetize all values added to the taxonomy
+                                                    groups[categoryIndex].children[typeIndex].children.push(subTypeObject);
+                                                    groups[categoryIndex].children[typeIndex].children.sort(this.compareValues('name'));
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Is this sorting the children each time a child is added?  Should we just sort the array once at the end after all of the children are added?

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 5 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Category is required, type is not. I will make this change.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 13 above)
Comment (ThomasSchellenbergNextCentury) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: And also with the sorting below (line 476).

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts [hunk 14]
```diff
@@ -0,0 +1,690 @@
...
+    onQuerySuccess(response: any) {
+        let groups = [];
+        let counter = 0;
+        this.taxonomyGroups = [];
+
+        try {
+            if (response && response.data && response.data.length && response.data[0]) {
+                this.isLoading = true;
+
+                response.data.forEach((d) => {
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts:373]: This whole loop is fairly hefty.  Could we pull parts of it out into smaller separate function(s)?

File: src/app/components/base-neon-component/base-neon.component.ts (hunk 1 above)
Comment (gitnordt) [reply src/app/components/base-neon-component/base-neon.component.ts]: Actually, that may be a better place for it. I didn't realize that this file housed utility functions. Good to know! I will move it and add a test.

File: src/app/components/network-graph/network-graph.component.ts (hunk 2 above)
Comment (gitnordt) [reply src/app/components/network-graph/network-graph.component.ts]: Thanks! I need to figure out how to turn the indentation formatting on automatically.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.html (hunk 3 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.html]: Done. THOR-908

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 6 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Done.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 7 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Yes, it is needed to capture the double click event defined in the taxonomy options. Without it, the double click event does not work. I will add a more detailed comment.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 8 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Done. AIDA-390

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 9 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Done. AIDA-390

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 11 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Done

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 12 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Fixed

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 10 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Done. AIDA-403

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.html (hunk 4 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.html:70]: ignoreSelf is not an option that should be provided as it deletes items in the taxonomy and I can't think of a reason why this would be the desired effect. Added ascending, idField, and filterFields.
Removing linkField for now since nothing has been implemented for it.

File: src/app/components/network-graph/network-graph.component.spec.ts [hunk 15]
```diff
@@ -457,16 +457,15 @@ describe('Component: NetworkGraph', () => {
             }]
         });
 
-        /*TODO:Broken in code so test is also broken
-        component.onSelect({nodes: ["testNodeValue2"]});
+        component.onSelect({nodes: ['testNodeValue2']});
```
Comment (bdpham) [inline src/app/components/network-graph/network-graph.component.spec.ts:460]: This test seems to be failing.

File: src/app/components/base-neon-component/base-neon.component.ts (hunk 1 above)
Comment (gitnordt) [reply src/app/components/base-neon-component/base-neon.component.ts]: Looking at this code again, I realized that this function is not really useful without sort() being applied so I am wrapping this code inside a sort().

File: src/app/components/network-graph/network-graph.component.spec.ts (hunk 15 above)
Comment (gitnordt) [reply src/app/components/network-graph/network-graph.component.spec.ts:460]: Fixed. Needed the multiFilterOperator to be set.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 13 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts]: Sorting now occurs after children are added.

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 14 above)
Comment (gitnordt) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts:373]: I've created a couple of new functions to handle some of the repetitive code, but I am not entirely sure if this is what you were looking for. Maybe we can move forward with this for now and if I or you think of more ways to condense the code that can be addressed at a later date.

File: src/app/components/network-graph/network-graph.component.ts (hunk 2 above)
Comment (ThomasSchellenbergNextCentury) [reply src/app/components/network-graph/network-graph.component.ts]: Hmm, it should work if you use `ng run lint`

File: src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts (hunk 14 above)
Comment (ThomasSchellenbergNextCentury) [reply src/app/components/taxonomy-viewer/taxonomy-viewer.component.ts:373]: This is good for now.  Thanks!

**Review by ThomasSchellenbergNextCentury — APPROVED** [review]
